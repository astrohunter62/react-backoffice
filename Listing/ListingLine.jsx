import React from 'react'
import PropTypes from 'prop-types'
import {
  TableCell,
  TableRow,
} from 'material-ui/Table'
import Checkbox from 'material-ui/Checkbox'


class ListingLine extends React.Component {
  static propTypes = {
    headers: PropTypes.arrayOf(PropTypes.object).isRequired,
    data: PropTypes.objectOf(PropTypes.any).isRequired,
    handleClick: PropTypes.func.isRequired,
    isSelected: PropTypes.bool.isRequired,
    handleKeyDown: PropTypes.func.isRequired,
    handleCheckClick: PropTypes.func.isRequired,
  }

  renderCells() {
    const { headers, data, handleClick } = this.props

    return headers.map((header) => {
      const content = data[header.id]
      let { transformContent } = header

      if (typeof transformContent !== 'function') {
        transformContent = contentData => contentData
      }

      return (
        <TableCell
          key={`header-${header.id}`}
          padding={header.disablePadding ? 'none' : 'default'}
          numeric={header.numeric}
          onClick={() => handleClick(data.id)}
        >
          {transformContent(content, data)}
        </TableCell>
      )
    })
  }

  render() {
    const {
      data,
      isSelected,
      handleKeyDown,
      handleCheckClick,
    } = this.props

    return (
      <TableRow
        hover
        onKeyDown={event => handleKeyDown(event, data.id)}
        role="checkbox"
        aria-checked={isSelected}
        tabIndex={-1}
        selected={isSelected}
      >
        <TableCell
          padding="checkbox"
          onClick={() => handleCheckClick(data.id)}
        >
          <Checkbox checked={isSelected} />
        </TableCell>

        {this.renderCells()}
      </TableRow>
    )
  }
}

export default ListingLine
