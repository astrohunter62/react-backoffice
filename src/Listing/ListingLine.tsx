import React from "react";
import { Checkbox, TableCell, TableRow } from "@material-ui/core";
import replace from "../utils/replace";
import { Header } from "./Listing";

const getCellContent = (content: any, transformContent: any, data: any) => {
  let printableContent = content;

  if (content?.highlight) {
    printableContent = content.value;
  }

  if (typeof transformContent === "function") {
    printableContent = transformContent(printableContent, data);
  }

  const props: Record<string, any> = {};

  if (typeof printableContent === "string" && content.highlight) {
    printableContent = replace(
      printableContent,
      content.highlight,
      (key: any) => `<span style="background-color: yellow;">${key}</span>`,
    );

    props.dangerouslySetInnerHTML = {
      __html: printableContent,
    };

    printableContent = undefined;
  }

  return {
    content: printableContent,
    props,
  };
};

type ListingLineProps = {
  headers: Header[];
  data: {
    [key: string]: any;
  };
  onClick?: (...args: any[]) => any;
  isSelected: boolean;
  handleKeyDown: (...args: any[]) => any;
  handleCheckClick: (...args: any[]) => any;
};

class ListingLine extends React.Component<ListingLineProps, {}> {
  renderCells() {
    const { headers, data, onClick } = this.props;

    return headers.map((header, index: number) => {
      const { content, props } = getCellContent(
        data[header.id],
        header.transformContent,
        data,
      );

      return (
        <TableCell
          key={`header-${header.id}-${index}`}
          padding={header.isPaddingDisabled ? "none" : "default"}
          align={header.isNumeric ? "right" : undefined}
          onClick={() => onClick && onClick(data.id)}
          {...props}
        >
          {content}
        </TableCell>
      );
    });
  }

  render() {
    const { data, isSelected, handleKeyDown, handleCheckClick } = this.props;

    return (
      <TableRow
        hover
        onKeyDown={(event) => handleKeyDown(event, data.id)}
        role="checkbox"
        aria-checked={isSelected}
        tabIndex={-1}
        selected={isSelected}
      >
        <TableCell padding="checkbox" onClick={() => handleCheckClick(data.id)}>
          <Checkbox color="primary" checked={isSelected} />
        </TableCell>

        {this.renderCells()}
      </TableRow>
    );
  }
}

export default ListingLine;
