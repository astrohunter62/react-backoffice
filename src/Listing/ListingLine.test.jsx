import React from 'react'
import renderer from 'react-test-renderer'
import { Table } from 'material-ui'

import ListingLine from './ListingLine'

import headers from '../../__tests__/data/listing_headers'

const twoHeaders = headers.slice(0, 2)

it('renders correctly', () => {
  const tree = renderer
    .create((
      <Table>
        <ListingLine
          headers={twoHeaders}
          data={{
            id: 1,
            name: 'Leanne Graham',
            username: 'Bret',
          }}
          onClick={() => { }}
          handleKeyDown={() => { }}
          handleCheckClick={() => { }}
          isSelected={false}
        />
      </Table>
    ))
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders selected', () => {
  const tree = renderer
    .create((
      <Table>
        <ListingLine
          headers={twoHeaders}
          data={{
            id: 1,
            name: 'Leanne Graham',
            username: 'Bret',
          }}
          onClick={() => { }}
          handleKeyDown={() => { }}
          handleCheckClick={() => { }}
          isSelected
        />
      </Table>
    ))
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders highlighted', () => {
  const tree = renderer
    .create((
      <Table>
        <ListingLine
          headers={twoHeaders}
          data={{
            id: 1,
            name: {
              value: 'Leanne Graham',
              highlight: 'Lea',
            },
            username: {
              value: 'Leanne Graham',
              highlight: 'Bret',
            },
          }}
          onClick={() => { }}
          handleKeyDown={() => { }}
          handleCheckClick={() => { }}
          isSelected
        />
      </Table>
    ))
    .toJSON()
  expect(tree).toMatchSnapshot()
})
