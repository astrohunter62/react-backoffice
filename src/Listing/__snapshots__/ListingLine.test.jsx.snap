// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<table
  className="MuiTable-root-1"
>
  <tr
    aria-checked={false}
    className="MuiTableRow-root-2 MuiTableRow-hover-6"
    onKeyDown={[Function]}
    role="checkbox"
    tabIndex={-1}
  >
    <td
      aria-sort={null}
      className="MuiTableCell-root-7 MuiTableCell-paddingCheckbox-13"
      onClick={[Function]}
    >
      <span
        className="MuiButtonBase-root-31 MuiIconButton-root-25 MuiSwitchBase-root-20 MuiSwitchBase-default-22 MuiCheckbox-default-15"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={null}
      >
        <span
          className="MuiIconButton-label-30"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-34"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
          <input
            checked={false}
            className="MuiSwitchBase-input-21"
            onChange={[Function]}
            type="checkbox"
          />
        </span>
        <span
          className="MuiTouchRipple-root-40"
        />
      </span>
    </td>
    <td
      aria-sort={null}
      className="MuiTableCell-root-7 MuiTableCell-paddingNone-14"
      onClick={[Function]}
    >
      Leanne Graham
    </td>
    <td
      aria-sort={null}
      className="MuiTableCell-root-7 MuiTableCell-paddingNone-14"
      onClick={[Function]}
    >
      Bret
    </td>
  </tr>
</table>
`;

exports[`renders highlighted 1`] = `
<table
  className="MuiTable-root-1"
>
  <tr
    aria-checked={true}
    className="MuiTableRow-root-2 MuiTableRow-hover-6 MuiTableRow-selected-5"
    onKeyDown={[Function]}
    role="checkbox"
    tabIndex={-1}
  >
    <td
      aria-sort={null}
      className="MuiTableCell-root-7 MuiTableCell-paddingCheckbox-13"
      onClick={[Function]}
    >
      <span
        className="MuiButtonBase-root-31 MuiIconButton-root-25 MuiSwitchBase-root-20 MuiSwitchBase-default-22 MuiCheckbox-default-15 MuiSwitchBase-checked-23 MuiCheckbox-checked-16 MuiCheckbox-checkedPrimary-17"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={null}
      >
        <span
          className="MuiIconButton-label-30"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-34"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
            />
          </svg>
          <input
            checked={true}
            className="MuiSwitchBase-input-21"
            onChange={[Function]}
            type="checkbox"
          />
        </span>
        <span
          className="MuiTouchRipple-root-40"
        />
      </span>
    </td>
    <td
      aria-sort={null}
      className="MuiTableCell-root-7 MuiTableCell-paddingNone-14"
      dangerouslySetInnerHTML={
        Object {
          "__html": "<span style=\\"background-color: yellow;\\">Lea</span>nne Graham",
        }
      }
      onClick={[Function]}
    />
    <td
      aria-sort={null}
      className="MuiTableCell-root-7 MuiTableCell-paddingNone-14"
      dangerouslySetInnerHTML={
        Object {
          "__html": "Leanne Graham",
        }
      }
      onClick={[Function]}
    />
  </tr>
</table>
`;

exports[`renders selected 1`] = `
<table
  className="MuiTable-root-1"
>
  <tr
    aria-checked={true}
    className="MuiTableRow-root-2 MuiTableRow-hover-6 MuiTableRow-selected-5"
    onKeyDown={[Function]}
    role="checkbox"
    tabIndex={-1}
  >
    <td
      aria-sort={null}
      className="MuiTableCell-root-7 MuiTableCell-paddingCheckbox-13"
      onClick={[Function]}
    >
      <span
        className="MuiButtonBase-root-31 MuiIconButton-root-25 MuiSwitchBase-root-20 MuiSwitchBase-default-22 MuiCheckbox-default-15 MuiSwitchBase-checked-23 MuiCheckbox-checked-16 MuiCheckbox-checkedPrimary-17"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={null}
      >
        <span
          className="MuiIconButton-label-30"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-34"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
            />
          </svg>
          <input
            checked={true}
            className="MuiSwitchBase-input-21"
            onChange={[Function]}
            type="checkbox"
          />
        </span>
        <span
          className="MuiTouchRipple-root-40"
        />
      </span>
    </td>
    <td
      aria-sort={null}
      className="MuiTableCell-root-7 MuiTableCell-paddingNone-14"
      onClick={[Function]}
    >
      Leanne Graham
    </td>
    <td
      aria-sort={null}
      className="MuiTableCell-root-7 MuiTableCell-paddingNone-14"
      onClick={[Function]}
    >
      Bret
    </td>
  </tr>
</table>
`;
